cpp/src/arrow/compute/exec/hash_join_benchmark.cc

```diff
@@ -0,0 +1,425 @@
+
+    auto l_schema = *l_schema_builder.Finish();
+    auto r_schema = *r_schema_builder.Finish();
+
+    l_batches_ =
+        MakeRandomBatches(l_schema, settings.num_probe_batches, settings.batch_size);
+    r_batches_ =
+        MakeRandomBatches(r_schema, settings.num_build_batches, settings.batch_size);
+
+    stats_.num_probe_rows = settings.num_probe_batches * settings.batch_size;
+
+    ctx_ = arrow::internal::make_unique<ExecContext>(
+        default_memory_pool(),
+        is_parallel ? arrow::internal::GetCpuThreadPool() : nullptr);
+
+    schema_mgr_ = arrow::internal::make_unique<HashJoinSchema>();
+    Expression filter = literal(true);
+    DCHECK_OK(schema_mgr_->Init(settings.join_type, *l_batches_.schema, left_keys,
+                                *r_batches_.schema, right_keys, filter, "l_", "r_"));
+
+    join_ = *HashJoinImpl::MakeBasic();
+
+    omp_set_num_threads(settings.num_threads);
+    auto schedule_callback = [](std::function<Status(size_t)> func) -> Status {
+#pragma omp task
+      { DCHECK_OK(func(omp_get_thread_num())); }
+      return Status::OK();
+    };
+
+    DCHECK_OK(join_->Init(
+        ctx_.get(), settings.join_type, !is_parallel /* use_sync_execution*/,
+        settings.num_threads, schema_mgr_.get(), {JoinKeyCmp::EQ}, std::move(filter),
+        [](ExecBatch) {}, [](int64_t x) {}, schedule_callback));
+  }
+
+  void RunJoin() {
+    double nanos = 0;
+#pragma omp parallel reduction(+ : nanos)
+    {
+      auto start = std::chrono::high_resolution_clock::now();
+      int tid = omp_get_thread_num();
+#pragma omp for nowait
+      for (auto batch : r_batches_.batches)
+        DCHECK_OK(join_->InputReceived(tid, 1 /* side */, batch));
+#pragma omp for nowait
+      for (auto batch : l_batches_.batches)
+        DCHECK_OK(join_->InputReceived(tid, 0 /* side */, batch));
+
+#pragma omp barrier
+
+#pragma omp single nowait
+      { DCHECK_OK(join_->InputFinished(tid, /* side */ 1)); }
+
+#pragma omp single nowait
+      { DCHECK_OK(join_->InputFinished(tid, /* side */ 0)); }
+      std::chrono::duration<double, std::nano> elapsed =
+          std::chrono::high_resolution_clock::now() - start;
+      nanos += elapsed.count();
+    }
+    stats_.total_nanoseconds = nanos;
+  }
+
+  ThreadIndexer thread_indexer_;
+  BatchesWithSchema l_batches_;
+  BatchesWithSchema r_batches_;
+  std::unique_ptr<HashJoinSchema> schema_mgr_;
+  std::unique_ptr<HashJoinImpl> join_;
+  std::unique_ptr<ExecContext> ctx_;
+
+  struct {
+    double total_nanoseconds;
+    uint64_t num_probe_rows;
+  } stats_;
+};
+
+static void HashJoinBasicBenchmarkImpl(benchmark::State& st,
+                                       BenchmarkSettings& settings) {
+  JoinBenchmark bm(settings);
+  double total_nanos = 0;
+  uint64_t total_rows = 0;
+  for (auto _ : st) {
+    bm.RunJoin();
+    total_nanos += bm.stats_.total_nanoseconds;
+    total_rows += bm.stats_.num_probe_rows;
+  }
+  st.counters["ns/row"] = total_nanos / total_rows;
```

Comment:
The thing is we're summing the time across all of the threads. I think settings it as a Rate will only do the total time. We actually want some measure of "efficiency". Multithreading is the primary such case, where we measure something like 70 ns/row for single-threaded but something like 120ns/row or more for higher numbers of threads. We want a way to see this overhead, which an overall rate wouldn't capture as well.

Reply:
I agree you'd want a cost that is summed across all threads.  [`MeasureProcessCPUTime`](https://github.com/google/benchmark/blob/main/docs/user_guide.md#cpu-timers) should address that.  However, if that doesn't play nicely with OpenMP for some reason then this approach is fine.

I don't see why this requires a time instead of a rate.  8.3M rows/s is still clearly worse than 14.3M rows/s

Reply:
Fair enough - I've changed that. I'll also ask @michalursa to weigh in on this. We've always measured in `cycles/row`, but there's not a good cross-platform way to measure cycles so we went with ns originally. Maybe rows/sec is fine.